### PR TITLE
[ONNX] Partially support tensor lists in loop/concat/stack

### PR DESCRIPTION
**Pull request by BowenBao**
This is a follow-up PR after https://github.com/pytorch/pytorch/pull/29136 ~~and https://github.com/pytorch/pytorch/pull/29171~~

ONNX::Loop does not support Sequence type as loop-carried dependencies. Only tensors are supported.
This PR adds a pass that converts Sequence loop-carried dependencies to scan_outputs.
In opset 11, only the below pattern is supported.
```
PTIR graph:
 ...
 %res.1 : Tensor[] = prim::ListConstruct()
 %res : Tensor[] = prim::Loop(%11, %22, %res.1)
   block0(%i.1 : Tensor, %res.6 : Tensor[]):
     ...
     %res.3 : Tensor[] = aten::append(%res.6, %17)
     -> (%22, %res.3)
 return (%res.3)

ONNX graph:
 ...
 %res : Tensor = onnx::Loop(%11, %22)
   block0(%i.1 : Tensor):
     ...
     -> (%22, %17)
 %res_seq : Tensor[] = onnx::SplitToSequence[keepdims=0](%res)
 return (%res_seq)
```

### CODE REVIEW

**Comment by neginraoof** [comment]
There is a seg fault at test/onnx/test_operators.py::TestOperators::test_arange_dynamic

**Comment by BowenBao** [comment]
> There is a seg fault at test/onnx/test_operators.py::TestOperators::test_arange_dynamic

Thanks, i'm looking into it now.

**Comment by BowenBao** [comment]
> There is a seg fault at test/onnx/test_operators.py::TestOperators::test_arange_dynamic

Thanks for catching this! This should be resolved now, re-running CI.

File: torch/csrc/jit/passes/onnx/prepare_inplace_ops_for_onnx.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      it->output()->copyMetadata(it->inputs()[0]);
+      it->inputs()[0]->replaceAllUsesAfterNodeWith(*it, it->output());
+    }
+  }
+}
+
+static void PrepareListAppendInsertForONNX(Block *b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      PrepareListPopForONNX(child_block);
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/prepare_inplace_ops_for_onnx.cpp]: Do you mean PrepareListAppendForONNX here?

File: torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp [hunk 2]
```diff
@@ -68,8 +70,172 @@ void FixupONNXLoops(Block* block) {
...
+  }
+
+  if (out_node->input(0) != in_val) {
+    // Only SequenceInsert that applies on loop-carried sequence is supported.
+    return false;
+  }
+
+  const auto seq_node_kind = loop_node->inputs()[i]->node()->kind();
+  if (seq_node_kind != ::c10::onnx::SequenceEmpty) {
+    // Initial sequence must be empty.
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: I'm just curious why the sequence has to be empty?

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: nvm

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -1491,6 +1502,112 @@ def forward(self, input):
...
+            def forward(self, x):
+                res = []
+                res1 = []
+                arr = x.split([3, 4, 1, 1, 2, 3, 2], 0)
+                res2 = torch.zeros(3, 4, dtype=torch.long)
+                for i in range(len(arr)):
+                    res = res.append(arr[i].sum(0, False))
+                    res1 = res1.append(arr[-1 - i].sum(0, False))
+                    res2 += 1
+                return torch.stack(res), torch.stack(res1), res2
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: Can we also add a pop in the loop test?
Thanks for making this work!

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: With the current ONNX spec, only single append is supported in loop.

File: torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp (hunk 2 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: The way we support append in loop is by converting the tensor list to scan output in ONNX loop. That requires the intial tensor list to be empty.

File: torch/csrc/jit/passes/onnx/prepare_inplace_ops_for_onnx.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/prepare_inplace_ops_for_onnx.cpp]: Changing to `PrepareListAppendAndInsertForONNX` ...

**Comment by BowenBao** [comment]
Looking at failed test `test_constant_fold_slice_negative_index`, it should be unrelated to this PR according to https://github.com/pytorch/pytorch/issues/30764

File: torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp [hunk 4]
```diff
@@ -68,8 +70,172 @@ void FixupONNXLoops(Block* block) {
...
+
+      concat_node->output()->replaceAllUsesWith(split_node->input());
+    }
+  }
+}
+
 void FixupONNXLoops(std::shared_ptr<Graph>& graph) {
   FixupONNXLoops(graph->block());
+  ConvertSequenceDependencies(graph->block());
+  fuseSequenceSplitConcat(graph->block());
```
Comment (lara-hdr) [inline torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: nit: start function name with upper case for consistency

File: torch/onnx/symbolic_registry.py [hunk 5]
```diff
@@ -26,6 +26,8 @@ def register_version(domain, version):
 def register_ops_helper(domain, version, iter_version):
     version_ops = get_ops_in_version(iter_version)
     for op in version_ops:
+        if op[0] == '_len':
```
Comment (lara-hdr) [inline torch/onnx/symbolic_registry.py]: why do we do this?

Comment (lara-hdr) [reply torch/onnx/symbolic_registry.py]: (nit : if there are more than one similar case we should eventually add them in a separate function that we call here)

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -1557,6 +1568,112 @@ def forward(self, input):
...
+                for i in range(x.size(2)):
+                    x = x + i
+                return x
+
+        model = LoopModel()
+        inputs = torch.zeros(1, 2, 3, dtype=torch.long)
+        self.run_test(model, inputs)
+
+    @skipIfUnsupportedMinOpsetVersion(9)
+    def test_loop_nested(self):
```
Comment (lara-hdr) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: does this work in tracing as well?

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: No, loop is not supported in tracing.

File: torch/onnx/symbolic_registry.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/symbolic_registry.py]: Agreed.. So far this is the only case that i could find, so I'm treating it as a special case handling.

File: torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp (hunk 4 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: thanks

**Review by lara-hdr — COMMENTED** [review]
Looks good, I have some minor comments

**Review by lara-hdr — APPROVED** [review]

**Review by houseroad — APPROVED** [review]
LGTM

File: torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp [hunk 7]
```diff
@@ -68,8 +70,172 @@ void FixupONNXLoops(Block* block) {
...
+//  return (%res_seq)
+void ConvertSequenceDependencies(Block* block) {
+  for (auto* node : block->nodes()) {
+    for (Block* block : node->blocks()) {
+      ConvertSequenceDependencies(block);
+    }
+
+    if (node->kind() == ::c10::onnx::Loop) {
+      auto* loop_node = node;
+      auto* graph = loop_node->owningGraph();
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: graph is not used.

File: torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp [hunk 8]
```diff
@@ -68,8 +70,172 @@ void FixupONNXLoops(Block* block) {
...
+//  ...
+//  %res : Tensor = onnx::Loop(%11, %22)
+//    block0(%i.1 : Tensor):
+//      ...
+//      -> (%22, %17)
+//  %res_seq : Tensor[] = onnx::SplitToSequence[keepdims=0](%res)
+//  return (%res_seq)
+void ConvertSequenceDependencies(Block* block) {
+  for (auto* node : block->nodes()) {
+    for (Block* block : node->blocks()) {
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp]: block is shadowed here.

**Pull request closed without merging**